storage/src/adb/append.rs

```diff
@@ -0,0 +1,884 @@
+}
+
+/// An append-only key-value ADB based on an MMR over its log of operations, supporting
+/// authentication of key-value pairs.
+pub struct Append<E: RStorage + Clock + Metrics, K: Array, V: Codec, H: CHasher, T: Translator> {
+    /// An MMR over digests of the operations applied to the db.
+    ///
+    /// # Invariant
+    ///
+    /// The number of leaves in this MMR always equals the number of operations in the unpruned
+    /// `log` and `locations_map`.
+    mmr: Mmr<E, H>,
+
+    /// A log of all operations applied to the db in order of occurrence. The _location_ of an
+    /// operation is its order of occurrence with respect to this log, and corresponds to its leaf
+    /// number in the MMR.
+    log: VJournal<E, Variable<K, V>>,
+
+    /// The number of operations that have been appended to the log (which must equal the number of
+    /// leaves in the MMR).
+    log_size: u64,
+
+    /// The number of items to put in each section of the journal.
+    log_items_per_section: u64,
+
+    /// A fixed-length journal that maps an operation's location to its offset within its respective
+    /// section of the log. (The section number is derived from location.)
+    locations_map: FJournal<E, U32>,
+
+    /// A map from each active key to the location of the operation that set its value.
+    ///
+    /// # Invariant
+    ///
+    /// Only references operations of type [Variable::Set].
+    snapshot: Index<T, u64>,
+
+    /// Cryptographic hasher to re-use within mutable operations requiring digest computation.
+    hasher: Standard<H>,
+}
+
+impl<E: RStorage + Clock + Metrics, K: Array, V: Codec, H: CHasher, T: Translator>
+    Append<E, K, V, H, T>
+{
+    /// Returns an [Append] adb initialized from `cfg`. Any uncommitted log operations will be
+    /// discarded and the state of the db will be as of the last committed operation.
+    pub async fn init(
+        context: E,
+        cfg: Config<T, <Variable<K, V> as Read>::Cfg>,
+    ) -> Result<Self, Error> {
+        let mut hasher = Standard::<H>::new();
+
+        let mut mmr = Mmr::init(
+            context.with_label("mmr"),
+            &mut hasher,
+            MmrConfig {
+                journal_partition: cfg.mmr_journal_partition,
+                metadata_partition: cfg.mmr_metadata_partition,
+                items_per_blob: cfg.mmr_items_per_blob,
+                write_buffer: cfg.mmr_write_buffer,
+                thread_pool: cfg.thread_pool,
+                buffer_pool: cfg.buffer_pool.clone(),
+            },
+        )
+        .await?;
+
+        let mut log = VJournal::init(
+            context.with_label("log"),
+            VConfig {
+                partition: cfg.log_journal_partition,
+                compression: cfg.log_compression,
+                codec_config: cfg.log_codec_config,
+                write_buffer: cfg.log_write_buffer,
+            },
+        )
+        .await?;
+
+        let mut locations_map = FJournal::init(
```

Comment:
ahhhhh, I meant `locations:` not `locations_map:` (I think it'll be less confusing as just `locations` given it isn't technically a `map`)

Reply:
lol, should have known (you hate multi-word names).  fixed.